arch/arm/core/aarch64/reset.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_internal.h>
+#include "vector_table.h"
+
+void __weak z_arm64_el_highest_plat_init(void)
+{
+	/* do nothing */
+}
+
+void __weak z_arm64_el3_plat_init(void)
+{
+	/* do nothing */
+}
+
+void __weak z_arm64_el2_plat_init(void)
+{
+	/* do nothing */
+}
+
+void __weak z_arm64_el1_plat_init(void)
+{
+	/* do nothing */
+}
+
+void z_arm64_el_highest_init(void)
+{
+	write_cntfrq_el0(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+
+	z_arm64_el_highest_plat_init();
+
+	isb();
+}
+
+void z_arm64_el3_init(void)
+{
+	uint64_t reg;
+
+	/* Setup vector table */
+	write_vbar_el3((uint64_t)_vector_table);
+	isb();
+
+	reg = read_cptr_el3();
+	reg &= ~(CPTR_EL3_RES0 |	/* RES0 */
+		 CPTR_EL3_TTA_BIT |	/* Do not trap sysreg accesses */
+		 CPTR_EL3_TFP_BIT |	/* Do not trap SVE, SIMD and FP */
+		 CPTR_EL3_TCPAC_BIT);	/* Do not trap CPTR_EL2 / CPACR_EL1 accesses */
+	reg |= CPTR_EL3_EZ_BIT;		/* Do not trap SVE accesses */
+	write_cptr_el3(reg);
+
+	reg = read_scr_el3();
+	reg &= ~(SCR_NS_BIT |		/* XXX: Boot in secure mode */
+		 SCR_IRQ_BIT |		/* Do not route IRQs to EL3 */
+		 SCR_FIQ_BIT |		/* Do not route FIQs to EL3 */
+		 SCR_EA_BIT);		/* Do not route Serrors to EL3 */
+	reg |= (SCR_RES1 |		/* RES1 */
+		SCR_RW_BIT |		/* EL2 execution state is AArch64 */
+		SCR_ST_BIT |		/* Do not trap EL1 accesses to timer */
+		SCR_HCE_BIT |		/* Do not trap HVC */
+		SCR_SMD_BIT);		/* Do not trap SMC */
+
+	write_scr_el3(reg);
+
+	z_arm64_el3_plat_init();
+
+	isb();
+}
+
+void z_arm64_el2_init(void)
+{
+	uint64_t reg;
+
+	reg = read_sctlr_el2();
+	reg |= (SCTLR_EL2_RES1 |	/* RES1 */
+		SCTLR_I_BIT |		/* Enable i-cache */
+		SCTLR_SA_BIT);		/* Enable SP alignment check */
+	write_sctlr_el2(reg);
+
+	reg = read_hcr_el2();
+	reg |= HCR_RW_BIT;		/* EL1 Execution state is AArch64 */
+	write_hcr_el2(reg);
+
+	zero_cntvoff_el2();		/* Set 64-bit virtual timer offset to 0 */
+	zero_cnthctl_el2();
+	zero_cnthp_ctl_el2();
+
+	z_arm64_el2_plat_init();
+
+	isb();
+}
+
+void z_arm64_el1_init(void)
+{
+	uint64_t reg;
+
+	/* Setup vector table */
+	write_vbar_el1((uint64_t)_vector_table);
+	isb();
+
+	write_cpacr_el1(CPACR_EL1_FPEN_NOTRAP); /* Do not trap NEON/SIMD/FP */
+
+	reg = read_sctlr_el1();
+	reg |= (SCTLR_EL1_RES1 |		/* RES1 */
+		SCTLR_I_BIT |			/* Enable i-cache */
+		SCTLR_SA_BIT);			/* Enable SP alignment check */
+	write_sctlr_el1(reg);
+
+	z_arm64_el1_plat_init();
+
+	isb();
+}
+
+void z_arm64_el3_get_next_el(uint64_t switch_addr)
+{
+	uint64_t spsr;
+
+	write_elr_el3(switch_addr);
+
+	/* Mask the DAIF */
+	spsr = SPSR_DAIF_MASK;
+
+	/*
+	 * Is considered an illegal return "[..] a return to EL2 when EL3 is
+	 * implemented and the value of the SCR_EL3.NS bit is 0 if
+	 * ARMv8.4-SecEL2is not implemented" (D1.11.2 from ARM DDI 0487E.a)
+	 *
+	 * XXX: boot with SCR_EL3.NS == 0
+	 */
+	if (is_el_implemented(2) && is_el2_sec_supported()) {
```

Comment:
If I read the extensions correctly, I think the logic here all told should be (pardon terrible pseudocode):

```c
if (is_el_implemented(2) && (!we_want_secure_mode() || is_el2_sec_supported()) {
    return EL2;
} else {
    return EL1;
}
```
That is, roughly:

1. If EL2 isn't implemented, we want EL1.NS / S, as appropriate.
2. If EL2 is implemented and we want secure mode, but EL2 doesn't support secure mode, skip EL2 and drop to EL1.S
3. Else drop to EL2.NS / S, as appropriate.

Reply:
Agree but the problem with this is that we are currently always booting with `SCR_EL3.NS==0` so at the current state the only way to drop into `EL2` is when we do support `EL2.S`. This is a problem of course and the main reason I opened #32492.

Reply:
For the purposes of this PR...

```c
static inline bool we_want_secure_mode(void) {
    return !IS_ENABLED(CONFIG_ARMV8_A_NS);
}
```

?

Reply:
ok, counter-proposal: what about instead of relying on `CONFIG_ARMV8_A_NS` we directly read the value of `SCR_EL3.NS`?
```c
static inline bool we_are_in_secure_mode(void) {
    return (SCR_EL3.NS == 0);
}

if (is_el_implemented(2) && we_are_in_secure_mode() && is_el2_sec_supported()) {
    return EL2;
} else {
    return EL1;
}
```

Reply:
> what about instead of relying on CONFIG_ARMV8_A_NS we directly read the value of SCR_EL3.NS?

> On a Warm reset, this field resets to an architecturally UNKNOWN value

In particular in the bare-metal case.

If we're in EL3 we get to choose the value of `SCR_EL3.NS` as we want to determine if we want to drop down into secure or nonsecure mode. That's our decision to make and it's irrelevant what it was before we booted.

(If we're not in EL3 the decision has already been made for us and we can't change it in a non-platform-specific way...)

Reply:
> If we're in EL3 we get to choose the value of `SCR_EL3.NS` as we want to determine if we want to drop down into secure or nonsecure mode. That's our decision to make and it's irrelevant what it was before we booted.

Yes, not questioning that. We are going to set `SCR_EL3.NS` before this check.

My point is that right now `SCR_EL3.NS` and `CONFIG_ARMV8_A_NS` are not related to each other. So you can be in a state where `CONFIG_ARMV8_A_NS` is set but `SCR_EL3.NS` is still `==0` (see details in #32492). So, considering this doing the check only on `CONFIG_ARMV8_A_NS` can actually results in an illegal return.

Reply:
> My point is that right now SCR_EL3.NS and CONFIG_ARMV8_A_NS are not related to each other. 

Ah. To be clear: this would be in combination with setting `SCR_EL3.NS` to the correct value for the configuration of `CONFIG_ARMV8_A_NS`.

Reply:
> Ah. To be clear: this would be in combination with setting SCR_EL3.NS to the correct value for the configuration of CONFIG_ARMV8_A_NS.

Haaaaaaa, ok then :smile: Setting `SCR_EL3.NS` in accordance with `CONFIG_ARMV8_A_NS` was something I was leaving for another PR but at this point I'll try to address everything in this PR adding a new commit on top.

Reply:
Ah. I mean, if you want to just directly read `SCR_EL3.NS` in this PR and adjust it in said other PR later on I'm fine with that.

Reply:
Nah, I just realized that after all it's a tiny change. I'll fix it in the next push.